### PR TITLE
feat(ui): Refactor `<GlobalSelectionHeader>`

### PR DESCRIPTION
**Pull request by billyvg**
This fixes a number of bugs with Global Selection Header including:

* Loading and saving of last used projects/envs with local storage. 
* race conditions with enforcement of a single project and the child view being mounted (and thus firing api requests before enforcing a single project)
* lost project context in certain situations eg after deleting a release.

### CODE REVIEW

**Comment by billyvg** [comment]
@matejminar could you help me testreleases v2? I’ve done basic tests but not sure what edge cases to test.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx [hunk 1]
```diff
@@ -609,133 +331,144 @@ class GlobalSelectionHeader extends React.Component<Props, State> {
...
-      </Header>
+              </HeaderItemPosition>
+            </React.Fragment>
+          )}
+
+          {!showEnvironmentSelector && <HeaderItemPosition isSpacer />}
+          {!showDateSelector && <HeaderItemPosition isSpacer />}
+        </Header>
+
+        {isGlobalSelectionReady ? children : <PageContent />}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:465]: The header now blocks content from mounting until it is ready. Eg projects are loaded and single project is enforced if necessary. I’ve opted to not include a loader, but we could add one in here.

**Comment by matejminar** [comment]
@billyvg I tested locally and things are looking good 👍

One minor thing I don't love is that when you initially visit URL with selected project (or do a refresh) the project selector flickers 4 times: Loading... -> My Projects -> Loading... -> internal

Also, the fact that there is a flash of Loading... on every navigation.

Not sure if it's a big deal though.

**Comment by billyvg** [comment]
@matejminar Thanks for testing! I agree, the loading state annoying -- I would prefer to get this merged sooner rather than later, as it fixes some bugs, but will work on solving this and either update this PR or make a new one.

File: src/sentry/static/sentry/app/actionCreators/globalSelection.tsx [hunk 2]
```diff
@@ -63,9 +77,137 @@ export function resetGlobalSelection() {
...
+  shouldEnforceSingleProject: boolean;
+  /**
+   * If true, do not load from local storage
+   */
+  skipLastUsed?: boolean;
+  defaultSelection?: Partial<GlobalSelection>;
+  forceProject?: MinimalProject | null;
+};
+
+export function initializeUrlState({
```
Comment (billyvg) [inline src/sentry/static/sentry/app/actionCreators/globalSelection.tsx:99]: This logic was spread out among the store and the component, I've moved them all in here so that it's easier to reason about all of these state changes.

File: src/sentry/static/sentry/app/actionCreators/globalSelection.tsx [hunk 3]
```diff
@@ -80,9 +222,7 @@ export function updateProjects(
     return;
   }
 
-  if (!router) {
-    GlobalSelectionActions.updateProjects(projects);
-  }
+  GlobalSelectionActions.updateProjects(projects);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/actionCreators/globalSelection.tsx:225]: These action creators now *always* create actions to update the store as well as update the URL. Previously, it would update URL and the component would update store.

File: src/sentry/static/sentry/app/components/charts/chartZoom.jsx [hunk 4]
```diff
@@ -103,20 +105,22 @@ class ChartZoom extends React.Component {
...
     // URL parameters are changed
-    callIfFunction(this.props.onZoom, {
+    callIfFunction(onZoom, {
       period,
       start: startFormatted,
       end: endFormatted,
     });
 
     this.zooming = () => {
-      updateParams(
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/charts/chartZoom.jsx:113]: I may change these to be private - I don't think we want to expose these since they do not update the store.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx [hunk 5]
```diff
@@ -86,6 +82,16 @@ type Props = {
    */
   selection: GlobalSelection;
 
+  /**
+   * Custom default selection values (e.g. a different default period)
+   */
+  defaultSelection?: Partial<GlobalSelection>;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:88]: Used by Performance.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx [hunk 6]
```diff
@@ -122,11 +128,6 @@ type Props = {
    */
   showRelative?: boolean;
 
-  /**
-   * Allow user to clear the time range selection
-   */
-  allowClearTimeRange?: boolean;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:128]: Previously only used by "Performance" - is replaced with `defaultSelection`. This also fixes inconsistency with default dates in "Performance".  cc @dashed

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx [hunk 7]
```diff
@@ -528,46 +284,12 @@ class GlobalSelectionHeader extends React.Component<Props, State> {
...
-      projects,
-      project => project.isMember
-    );
-
-    if (isSuperuser || isOrgAdmin) {
-      return [memberProjects, nonMemberProjects];
-    }
-
-    return [memberProjects, []];
-  };
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:564]: Moved up a component level.

File: src/sentry/static/sentry/app/stores/globalSelectionStore.jsx [hunk 8]
```diff
@@ -1,143 +1,60 @@
...
+    this.selection = newSelection;
+    this.trigger(this.get());
   },
 
   get() {
-    return this.selection;
+    return {
+      selection: this.selection,
+      isReady: this.isReady(),
+    };
```
Comment (billyvg) [inline src/sentry/static/sentry/app/stores/globalSelectionStore.jsx:52]: This is safe-ish since nothing accesses the store directly (instead it's accessed via `withGlobalSelection` HoC).

File: src/sentry/static/sentry/app/views/events/index.jsx [hunk 9]
```diff
@@ -62,29 +62,29 @@ class EventsContainer extends React.Component {
...
-            <Body>
-              <PageHeader>
-                <HeaderTitle>
-                  {t('Events')} <BetaTag />
-                </HeaderTitle>
-                <StyledSearchBar
-                  organization={organization}
-                  projectIds={selection.projects}
-                  query={(location.query && location.query.query) || ''}
-                  placeholder={t('Search for events, users, tags, and everything else.')}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/events/index.jsx:80]: ???

File: src/sentry/static/sentry/app/views/organizationGroupDetails/groupDetails.tsx [hunk 10]
```diff
@@ -59,10 +60,14 @@ class GroupDetails extends React.Component<Props, State> {
...
 
-  componentDidUpdate(_prevProps, prevState: State) {
+  componentDidUpdate(prevProps: Props, prevState: State) {
     if (prevState.loading && !this.state.loading) {
       callIfFunction(this.props.finishProfile);
     }
+
+    if (prevProps.isGlobalSelectionReady !== this.props.isGlobalSelectionReady) {
+      this.fetchData();
+    }
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationGroupDetails/groupDetails.tsx:70]: This view will not make a request until store is ready, so we need to check for this update so that we can make the request.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx (hunk 6 above)
Comment (dashed) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:128]: Ok.

File: tests/js/spec/views/discover/index.spec.jsx [hunk 11]
```diff
@@ -63,26 +63,6 @@ describe('DiscoverContainer', function() {
...
     });
-
-    it('sets active projects from global selection', async function() {
-      ProjectsStore.loadInitialData(organization.projects);
-
-      GlobalSelectionStore.reset({
-        projects: [1],
-        environments: [],
-        datetime: {start: null, end: null, period: '14d'},
-      });
```
Comment (billyvg) [inline tests/js/spec/views/discover/index.spec.jsx:74]: Removing this test because we don't use the store like this.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/initializeGlobalSelectionHeader.tsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      shouldEnforceSingleProject,
+    } = this.props;
+
+    // Make an exception for issue details in the case where it is accessed directly (e.g. from email)
+    // We do not want to load the user's last used env/project in this case, otherwise will
+    // lead to very confusing behavior.
+    //
+    // `routes` is only ever undefined in tests
+    const skipLastUsed = !!routes?.find(
+      ({path}) => path && path.includes('/organizations/:orgId/issues/:groupId/')
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/initializeGlobalSelectionHeader.tsx:50]: Are there other properties we could add to routes so we're not matching on paths?

File: src/sentry/static/sentry/app/views/events/index.jsx [hunk 13]
```diff
@@ -62,29 +62,29 @@ class EventsContainer extends React.Component {
...
-                  onSearch={this.handleSearch}
-                />
-              </PageHeader>
-              {children}
-            </Body>
-          </LightWeightNoProjectMessage>
-        </PageContent>
+        <GlobalSelectionHeader resetParamsOnChange={['cursor']}>
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/events/index.jsx:65]: Having the selection header wrap the page is nice as it can throttle the rendering of the rest of the page. :+1:

File: src/sentry/static/sentry/app/components/charts/chartZoom.jsx (hunk 4 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/charts/chartZoom.jsx:113]: So I believe the issue I had with this is that echarts doesn't expose an event when the chart is finished zooming. I can look into it again and see if it has changed since, but would prefer that to be in a follow-up if it is possible

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx (hunk 5 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:88]: I think this warrants a bigger discussion with Product /cc @adhiraj?

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/initializeGlobalSelectionHeader.tsx (hunk 12 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/initializeGlobalSelectionHeader.tsx:50]: Hmm I don't remember why we didn't do that in the first place, but I'll give it a shot. Typescript might complain.

File: src/sentry/static/sentry/app/components/charts/chartZoom.jsx (hunk 4 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/charts/chartZoom.jsx:113]: So from what I'm seeing this is what is happening:

1) User zooms
2) eCharts starts zoom animation
3) eCharts finishes zoom animation
4) Update URL + store
5) Fetch new data (clears chart)
6) Renders new zoomed in chart based on new absolute date (which is shaped a bit differently than the eCharts local zoom chart shape).

Comment (markstory) [reply src/sentry/static/sentry/app/components/charts/chartZoom.jsx:113]: Yeah the animation and then reload is unpleasant. Fixing that separately totally makes sense as it is a tricky problem.

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx (hunk 5 above)
Comment (adhiraj) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:88]: @markstory What's your thinking? My gut is that most people are ok with (and perhaps prefer) 7d to 14d given sentry's natural usage frequency is weekly. I can do a quick pulse on the team channel too.
Also just to confirm, the default is used by issues, releases, and user feedback?

**Comment by billyvg** [comment]
@markstory anything else needed on this?

**Review by markstory — APPROVED** [review]

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx (hunk 5 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.tsx:88]: @adhiraj The default is used everywhere except Performance/Discover - so all the above mentioned + Events/Discover (which is going away soon) and Dashboards
